src/relay/transforms/compiler_function_utils.cc

```diff
@@ -0,0 +1,201 @@
+      : cache_(cache), compiler_filter_(std::move(compiler_filter)), mod_(std::move(mod)) {}
+
+  Expr Rewrite_(const CallNode* pre, const Expr& post) final {
+    Call new_call = Downcast<Call>(post);
+    if (const auto* function_node = new_call->op.as<FunctionNode>()) {
+      Optional<String> opt_compiler = function_node->GetAttr<String>(attr::kCompiler);
+      if (opt_compiler.defined() &&
+          (compiler_filter_.empty() || opt_compiler.value() == compiler_filter_)) {
+        auto function = GetRef<Function>(function_node);
+        ICHECK(FreeVars(function).empty()) << "Function marked with '" << attr::kCompiler
+                                           << "' attribute should not have free variables";
+        GlobalVar global_symbol = cache_->GetGlobalSymbol(function);
+        // Depending on the cache's implementation, two structurally equal (but not object equal)
+        // functions may be assigned the same global symbol. If so we'll lift it just once, but
+        // rewrite all the calls.
+        if (!mod_->ContainGlobalVar(global_symbol->name_hint)) {
+          function =
+              WithAttr(std::move(function), tvm::attr::kGlobalSymbol, global_symbol->name_hint);
+          mod_->Add(global_symbol, function);
+        }
+        return WithFields(new_call, global_symbol);
+      }
+    }
+    return post;
+  }
+
+ private:
+  GlobalSymbolCache* cache_;
+  std::string compiler_filter_;
+  IRModule mod_;
+};
+
+/*!
+ * \brief Rewrite calls to global "Compiler" functions to use the 'call_lowered' convention.
+ */
+class CallRewriter : public MixedModeMutator {
+ public:
+  CallRewriter(std::string compiler_filter, IRModule mod)
+      : compiler_filter_(std::move(compiler_filter)), mod_(std::move(mod)) {}
+
+  Expr Rewrite_(const CallNode* pre, const Expr& post) final {
+    Call new_call = Downcast<Call>(post);
+    if (const auto* global_var_node = new_call->op.as<GlobalVarNode>()) {
+      if (const auto* function_node =
+              mod_->Lookup(GetRef<GlobalVar>(global_var_node)).as<FunctionNode>()) {
+        Optional<String> opt_compiler = function_node->GetAttr<String>(attr::kCompiler);
+        if (opt_compiler.defined() &&
+            (compiler_filter_.empty() || opt_compiler.value() == compiler_filter_)) {
+          Optional<String> opt_global_symbol =
+              function_node->GetAttr<String>(tvm::attr::kGlobalSymbol);
+          ICHECK(opt_global_symbol.defined());
+          GlobalVar global_symbol = mod_->GetGlobalVar(opt_global_symbol.value());
+          CallLoweredAttrs attrs;
+          attrs.metadata.Set("relay_attrs", new_call->attrs);
+          return CallLowered(global_symbol, new_call->args, attrs, new_call->span);
+        }
+      }
+    }
+    return post;
+  }
+
+ private:
+  std::string compiler_filter_;
+  IRModule mod_;
+};
+
+}  // namespace
+
+GlobalVar ExistingGlobalSymbolCache::GetGlobalSymbol(const Function& function) {
```

Comment:
Currently yes. But Collage has another one since it is the one generating the global_symbols:
https://github.com/mbs-octoml/mbs-tvm/blob/908e8c9b0d4e09dede18e0a2db7f0d3220018c6f/src/relay/collage/candidate_function_cache.cc#L31